Optimize WTF::spanReinterpretCast

ddkilzer opened the pull request:
#### 9359aad17416457adec2639e7ef4acec30cbddfa
<pre>
Optimize WTF::spanReinterpretCast
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=280032">https://bugs.webkit.org/show_bug.cgi?id=280032</a>&gt;
&lt;<a href="https://rdar.apple.com/136330115">rdar://136330115</a>&gt;

Reviewed by Alex Christensen.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::spanReinterpretCast):
- Fix Extent in return type when std::dynamic_extent is not used.
- Prevent change of size in bytes when passing through
  spanReinterpretCast():
  - Limit RELEASE_ASSERT() to cases where std::dynamic_extent is used,
    and one of the following criteria:
    - The source type is smaller than the destination type, or
    - The source type is not an integral multiple of the destination
      type.
  - Use static_assert() when not using std::dynamic_extent.
- Fix name of function in message for static_assert() const-ness check.

* Tools/TestWebKitAPI/Tests/WTF/StdLibExtrasTests.cpp:
(TestWebKitAPI::TEST(WTF_StdLibExtras, SpanReinterpretCast_DynamicExtent)): Add.
(TestWebKitAPI::TEST(WTF_StdLibExtras, SpanReinterpretCast_DynamicExtent_ManualRuntimeErrors)): Add.
(TestWebKitAPI::TEST(WTF_StdLibExtras, SpanReinterpretCast_NonDynamicExtent)): Add.
(TestWebKitAPI::TEST(WTF_StdLibExtras, SpanReinterpretCast_NonDynamicExtent_CompileTimeErros)): Add.
- Add tests.

Canonical link: <a href="https://commits.webkit.org/284316@main">https://commits.webkit.org/284316@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/10318339a1209a49a34ee0ce0b1afe5bdf0489e9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68990 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48390 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21662 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73070 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20141 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71107 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56191 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19996 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54945 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13394 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72056 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44178 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59565 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35418 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40843 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16994 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18522 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/62106 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62792 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17340 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74778 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68236 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12972 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16585 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62589 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13010 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59648 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62486 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10463 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4073 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90017 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10545 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44194 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15960 "Found 59 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/es6/regex-quantifiers.js.default, ChakraCore.yaml/ChakraCore/test/es6/regex-set.js.default, jsc-layout-tests.yaml/js/script-tests/dfg-arguments-alias-one-block.js.layout, jsc-layout-tests.yaml/js/script-tests/dfg-arguments-osr-exit-multiple-blocks-before-exit.js.layout, jsc-layout-tests.yaml/js/script-tests/dfg-array-dead.js.layout, stress/arguments-non-configurable.js.bytecode-cache, stress/async-await-mozilla.js.bytecode-cache, stress/async-await-mozilla.js.dfg-eager, stress/big-int-bitwise-and-to-primitive-precedence.js.bytecode-cache ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45267 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46463 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45009 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -781,8 +781,16 @@ template<typename OptionalType> auto valueOrDefault(OptionalType&& optionalValue
+
+    static_assert(std::is_const_v<T> || (!std::is_const_v<T> && !std::is_const_v<U>), "spanReinterpretCast will not remove constness from source");
+
+    if constexpr (std::is_same_v<T, std::byte>)
```
I'm not sure these are worth the additional complexity for something that doesn't change anything.

achristensen07 replied:
Let's instead add static_assert(!std::is_same<std::remove_const_t<T>, std::byte>, "Use std::as_bytes or std::as_writable_bytes instead")

ddkilzer replied:
Via offline discussion, I decided to remove the `std::as_[writable_]bytes` early returns.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WTF/StdLibExtrasTests.cpp:
```diff
@@ -206,4 +207,19 @@ TEST(WTF_StdLibExtras, ByteCast)
+    auto signedIntSpan = signedInt.span();
+
+    auto signedIntByteSpan = spanReinterpretCast<const uint8_t>(signedIntSpan);
+    EXPECT_TRUE(!memcmp(signedIntSpan.data(), signedIntByteSpan.data(), signedIntByteSpan.size_bytes()));
```
I'm not sure this tests anything valuable because both data() results point to the same thing.
Could we add a test that makes a span of 2 things that are 24 bytes each and spanReinterpretCasts it to a span of three things that are 16 bytes each?

Now that I think about it, we may want to not allow such a reinterpreting, and we might just want to change the release assertion to static_assert(!(sizeof(U) % sizeof(T)))

ddkilzer replied:
> Now that I think about it, we may want to not allow such a reinterpreting, and we might just want to change the release assertion to `static_assert(!(sizeof(U) % sizeof(T)))`

We'd want to allow casting in both directions (large-to-small and small-to-large)—which is basically encoding to bytes and decoding from bytes—correct?

```
    static_cast((sizeof(U) == sizeof(T)) || (sizeof(U) > sizeof(T) && !(sizeof(U) % sizeof(T))) || (sizeof(U) < sizeof(T) && !(sizeof(T) % sizeof(U))));
```

ddkilzer replied:
I also agree we should do this, because if the user wants to shorten the `std::span`, they should do it _before_ passing it to `WTF::spanReinterpretCast<>()` so that `spanReinterpretCast` is doing one and only one thing which does not change the value of `size_bytes()`.

ddkilzer replied:
> Could we add a test that makes a span of 2 things that are 24 bytes each and spanReinterpretCasts it to a span of three things that are 16 bytes each?

You could always go from 2 things that are 24 bytes each -> 48 things of 1 byte each -> 3 things that are 16 bytes each—but you'd have to be very explicit about the casting.

ddkilzer replied:
> I'm not sure this tests anything valuable because both data() results point to the same thing.

Should I verify individual values coming out of the reinterpreted span object?

Maybe we only need to test one `spanReinterpretCast` call?  (Seemed silly not to have at least one test for it, though.)

ddkilzer replied:
Okay, I think I addressed this in the latest PR.

ddkilzer replied:
> Okay, I think I addressed this in the latest PR.

Well, not quite.

Casting from larger to smaller is always okay, but casting from smaller to larger requires knowing size_bytes(), so we still have to have the `RELEASE_ASSERT()` when going from larger to smaller and extent is set to `std::dynamic_extent`.

ddkilzer replied:
Fixed.

achristensen07 commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -779,11 +779,18 @@ template<typename OptionalType> auto valueOrDefault(OptionalType&& optionalValue
 
 template<typename T, typename U, std::size_t Extent>
-std::span<T, Extent> spanReinterpretCast(std::span<U, Extent> span)
+std::span<T, Extent == std::dynamic_extent ? std::dynamic_extent : (sizeof(U) * Extent) / sizeof(T)> spanReinterpretCast(std::span<U, Extent> span)
```
Would be nice to add a test or two with a non-dynamic extent.

achristensen07 reviewed (approved):
lgtm still.

Commit:
Optimize WTF::spanReinterpretCast
<https://bugs.webkit.org/show_bug.cgi?id=280032>
<rdar://136330115>

Reviewed by Alex Christensen.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::spanReinterpretCast):
- Fix Extent in return type when std::dynamic_extent is not used.
- Prevent change of size in bytes when passing through
  spanReinterpretCast():
  - Limit RELEASE_ASSERT() to cases where std::dynamic_extent is used,
    and one of the following criteria:
    - The source type is smaller than the destination type, or
    - The source type is not an integral multiple of the destination
      type.
  - Use static_assert() when not using std::dynamic_extent.
- Fix name of function in message for static_assert() const-ness check.

* Tools/TestWebKitAPI/Tests/WTF/StdLibExtrasTests.cpp:
(TestWebKitAPI::TEST(WTF_StdLibExtras, SpanReinterpretCast_DynamicExtent)): Add.
(TestWebKitAPI::TEST(WTF_StdLibExtras, SpanReinterpretCast_DynamicExtent_ManualRuntimeErrors)): Add.
(TestWebKitAPI::TEST(WTF_StdLibExtras, SpanReinterpretCast_NonDynamicExtent)): Add.
(TestWebKitAPI::TEST(WTF_StdLibExtras, SpanReinterpretCast_NonDynamicExtent_CompileTimeErros)): Add.
- Add tests.

Canonical link: https://commits.webkit.org/284316@main
